### PR TITLE
kubectl run to produce deployment and job

### LINKED ISSUE #14954: Kubectl run should produce deployment and job
Forked from #1743.

Kubectl run should produce deployment and job rather than replication controller and raw pods.

Initially using an experimental generator version, which we can make the default once these APIs reach beta+ maturity.

### PR DESCRIPTION
**Pull request by janetkuo**
Fixes #14954. 
- Implement generators for deployment (`deployment/v1beta1`) and job (`job/v1beta1`). 
- Make `kubectl run` default to produce rcs and jobs (instead of rcs and pods). Will replace rcs with deployments once Deployment reaches beta in the future (not in this PR). 
- If `--restart=Always`, a rc will be generated; otherwise a job will be generated. 

TODO:
- [x] finish job/deployment related flags and help text (for kubectl run cmd)
- [x] finish job/deployment fields (for generator)
- [x] add tests

@kubernetes/kubectl @nikhiljindal @ironcladlou

### CODE REVIEW

**Comment by kargakis** [comment]
cc: @soltysh

**Comment by janetkuo** [comment]
This PR is ready for review now. PR description is updated too. PTAL.

File: pkg/kubectl/cmd/run.go [hunk 1]
```diff
@@ -34,7 +35,7 @@ import (
 
 const (
 	run_long = `Create and run a particular image, possibly replicated.
-Creates a replication controller to manage the created container(s).`
+Creates a deployment or job to manage the created container(s).`
```
Comment (kargakis) [inline pkg/kubectl/cmd/run.go]: run still supports rcs, is this change intentional so to drive people away from rcs?

Comment (kargakis) [inline pkg/kubectl/cmd/run.go]: hm, I see you have removed the rc and pod generators below. Isn't this going to break people?

File: pkg/kubectl/cmd/run.go [hunk 2]
```diff
@@ -97,13 +101,15 @@ func addRunFlags(cmd *cobra.Command) {
...
 	cmd.Flags().Bool("command", false, "If true and extra arguments are present, use them as the 'command' field in the container, rather than the 'args' field which is the default.")
 	cmd.Flags().String("requests", "", "The resource requirement requests for this container.  For example, 'cpu=100m,memory=256Mi'")
 	cmd.Flags().String("limits", "", "The resource requirement limits for this container.  For example, 'cpu=200m,memory=512Mi'")
 	cmd.Flags().Bool("expose", false, "If true, a public, external service is created for the container(s) which are run")
 	cmd.Flags().String("service-generator", "service/v2", "The name of the generator to use for creating a service.  Only used if --expose is true")
 	cmd.Flags().String("service-overrides", "", "An inline JSON override for the generated service object. If this is non-empty, it is used to override the generated object. Requires that the object supply a valid apiVersion field.  Only used if --expose is true.")
+	cmd.Flags().IntP("parallel", "", -1, "The maximum desired number of pods the job should run at any given time. Default to the value of --complete.")
```
Comment (kargakis) [inline pkg/kubectl/cmd/run.go]: --parallel in oc run is --replicas in oc scale. We may want to stay consistent here?

Anyway why we decided in scaling parallelism vs complete @soltysh ?

File: pkg/kubectl/cmd/run.go [hunk 3]
```diff
@@ -124,6 +130,13 @@ func Run(f *cmdutil.Factory, cmdIn io.Reader, cmdOut, cmdErr io.Writer, cmd *cob
 	if interactive && replicas != 1 {
 		return cmdutil.UsageError(cmd, fmt.Sprintf("-i/--stdin requires that replicas is 1, found %d", replicas))
 	}
+	complete := cmdutil.GetFlagInt(cmd, "complete")
```
Comment (kargakis) [inline pkg/kubectl/cmd/run.go]: if you don't use complete anywhere else, inline in the if like parallel.

File: pkg/kubectl/cmd/run.go (hunk 1 above)
Comment (smarterclayton) [inline pkg/kubectl/cmd/run.go]: We cannot remove those without deprecating them and notifying users.  I
suspect we do not want to remove them before 1.2


On Nov 14, 2015, at 9:16 AM, Michail Kargakis <notifications@github.com>
wrote:

In pkg/kubectl/cmd/run.go
<https://github.com/kubernetes/kubernetes/pull/17195#discussion_r44858944>:

> @@ -34,7 +35,7 @@ import (
>
>  const (
>  	run_long = `Create and run a particular image, possibly replicated.
> -Creates a replication controller to manage the created container(s).`
> +Creates a deployment or job to manage the created container(s).`

hm, I see you have removed the rc and pod generators below. Isn't this
going to break people?

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/17195/files#r44858944>.

File: hack/test-cmd.sh [hunk 4]
```diff
@@ -647,6 +647,23 @@ runTests() {
...
+  # Post-Condition: Job "pi" is created 
+  kube::test::get_object_assert jobs "{{range.items}}{{$id_field}}:{{end}}" 'pi:'
+  # Pre-Condition: no Deployment is running 
+  kube::test::get_object_assert deployment "{{range.items}}{{$id_field}}:{{end}}" ''
+  # Command
+  kubectl run nginx --image=nginx 
+  # Post-Condition: Deployment "nginx" is created 
+  kube::test::get_object_assert deployment "{{range.items}}{{$id_field}}:{{end}}" 'nginx:'
+  # Clean up 
+  kubectl delete jobs pi
```
Comment (soltysh) [inline hack/test-cmd.sh]: I'd generally try to separate the tests, I'd suggest cleaning jobs before starting deployments.

File: pkg/kubectl/cmd/run.go (hunk 1 above)
Comment (soltysh) [inline pkg/kubectl/cmd/run.go]: I see all the previous generators in place (factory.go), just adding the new ones and convincing people to use the new ones. I'm perfectly ok with this approach.

File: pkg/kubectl/cmd/run.go (hunk 2 above)
Comment (soltysh) [inline pkg/kubectl/cmd/run.go]: > --parallel in oc run is --replicas in oc scale. We may want to stay consistent here?

I didn't want to introduce new flag, but after a while I see this bring confusion. I'll create a separate issue to discuss naming of that parameter in scale command.

Anyway why we decided in scaling parallelism vs complete @soltysh ?

File: pkg/kubectl/cmd/run.go (hunk 1 above)
Comment (kargakis) [inline pkg/kubectl/cmd/run.go]: The problem here is that we change the defaults. Someone previously running `kubectl run` w/o specifying a generator would get a rc running their image, now they will get a deployment. I am also in favor of switching to deployments but I am not sure of changing defaults.

File: pkg/kubectl/cmd/run.go (hunk 2 above)
Comment (soltysh) [inline pkg/kubectl/cmd/run.go]: See #17313.

Comment (soltysh) [inline pkg/kubectl/cmd/run.go]: > Anyway why we decided in scaling parallelism vs complete @soltysh ?

We've decided you want to have the completion number predictable when running a particular task, allowing user only to change the actual execution by changing up/down the parallelism parameter.

File: pkg/kubectl/cmd/run.go (hunk 3 above)
Comment (soltysh) [inline pkg/kubectl/cmd/run.go]: It's used in defaulting parallel if that is <= 0.

File: pkg/kubectl/cmd/run.go (hunk 1 above)
Comment (janetkuo) [inline pkg/kubectl/cmd/run.go]: Talked with @bgrant0607 about this. The users should specify a generator when they want to depend on a specific version of the generator (for example, `run/v1`). Since job is beta I'll keep it as default; but deployment isn't beta yet, so keep rc as default.

Comment (soltysh) [inline pkg/kubectl/cmd/run.go]: SGTM.

File: pkg/apis/extensions/validation/validation.go [hunk 5]
```diff
@@ -293,7 +293,10 @@ func ValidateDeploymentSpec(spec *extensions.DeploymentSpec) errs.ValidationErro
 	allErrs = append(allErrs, apivalidation.ValidatePositiveField(int64(spec.Replicas), "replicas")...)
 	allErrs = append(allErrs, apivalidation.ValidatePodTemplateSpecForRC(&spec.Template, spec.Selector, spec.Replicas, "template")...)
 	allErrs = append(allErrs, ValidateDeploymentStrategy(&spec.Strategy, "strategy")...)
-	allErrs = append(allErrs, apivalidation.ValidateLabelName(spec.UniqueLabelKey, "uniqueLabel")...)
+	// empty string is a valid UniqueLabelKey
```
Comment (bgrant0607) [inline pkg/apis/extensions/validation/validation.go]: Why is empty string valid? It should be set by default prior to validation.
cc @nikhiljindal

File: pkg/kubectl/cmd/run.go [hunk 6]
```diff
@@ -120,8 +124,16 @@ func Run(f *cmdutil.Factory, cmdIn io.Reader, cmdOut, cmdErr io.Writer, cmd *cob
...
+	if err != nil {
+		return err
+	}
+	ignoreReplicas := false
+	if restartPolicy == api.RestartPolicyNever || restartPolicy == api.RestartPolicyOnFailure {
+		ignoreReplicas = true
+	}
 	replicas := cmdutil.GetFlagInt(cmd, "replicas")
-	if interactive && replicas != 1 {
+	if interactive && replicas != 1 && !ignoreReplicas {
```
Comment (bgrant0607) [inline pkg/kubectl/cmd/run.go]: Please ensure that interactive mode still works with this change.

File: pkg/kubectl/cmd/run.go [hunk 7]
```diff
@@ -120,8 +124,16 @@ func Run(f *cmdutil.Factory, cmdIn io.Reader, cmdOut, cmdErr io.Writer, cmd *cob
 	if tty && !interactive {
 		return cmdutil.UsageError(cmd, "-i/--stdin is required for containers with --tty=true")
 	}
+	restartPolicy, err := getRestartPolicy(cmd, interactive)
+	if err != nil {
+		return err
+	}
+	ignoreReplicas := false
```
Comment (bgrant0607) [inline pkg/kubectl/cmd/run.go]: Why ignore replicas rather than just require replicas=1?

File: pkg/apis/extensions/validation/validation.go (hunk 5 above)
Comment (janetkuo) [inline pkg/apis/extensions/validation/validation.go]: From https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/extensions/v1beta1/types.go#L210-L211:
> Users can set this to an empty string to indicate that the system should not add any selector and label.

File: pkg/kubectl/cmd/run.go (hunk 2 above)
Comment (janetkuo) [inline pkg/kubectl/cmd/run.go]: Removed --parallel and --complete. We don't need to support this in `kubectl run`.

File: hack/test-cmd.sh (hunk 4 above)
Comment (janetkuo) [inline hack/test-cmd.sh]: Fixed

File: pkg/kubectl/cmd/run.go (hunk 7 above)
Comment (janetkuo) [inline pkg/kubectl/cmd/run.go]: Fixed

File: pkg/kubectl/cmd/run.go (hunk 2 above)
Comment (soltysh) [inline pkg/kubectl/cmd/run.go]: Why? I found specifying completions very useful here, you can't change it afterwards. Scale operates only on parallelism.

File: pkg/apis/extensions/validation/validation.go (hunk 5 above)
Comment (nikhiljindal) [inline pkg/apis/extensions/validation/validation.go]: API allows users to set it to empty string to indicate that no label should be added. The default value will be set if user does not set it, not if user sets it to empty string (its a string pointer)
From API comments:
>
        // Users can set this to an empty string to indicate that the system should
	// not add any selector and label. If unspecified, system uses
	// "deployment.kubernetes.io/podTemplateHash".
	// Value of this key is hash of DeploymentSpec.PodTemplateSpec.
	// No label is added if this is set to empty string.
>

**Comment by janetkuo** [comment]
All green now.

File: pkg/kubectl/cmd/run.go (hunk 2 above)
Comment (janetkuo) [inline pkg/kubectl/cmd/run.go]: Per discussion with @bgrant0607, we don't need to support everything of jobs in `kubectl run`. To not making it too complicated we don't put `--parallel` and `--complete` for now; we can add them later if we need to.

**Comment by soltysh** [comment]
LGTM, please squash your commits though.

**Comment by janetkuo** [comment]
Commits squashed.

File: pkg/kubectl/cmd/run.go [hunk 8]
```diff
@@ -83,7 +86,8 @@ func NewCmdRun(f *cmdutil.Factory, cmdIn io.Reader, cmdOut, cmdErr io.Writer) *c
 }
 
 func addRunFlags(cmd *cobra.Command) {
-	cmd.Flags().String("generator", "", "The name of the API generator to use.  Default is 'run/v1' if --restart=Always, otherwise the default is 'run-pod/v1'.")
+	// TODO: Change the default to "deployment/v1beta1" when deployment reaches beta (#15313)
```
Comment (kargakis) [inline pkg/kubectl/cmd/run.go]: Remove this TODO since the default should be set if restartPolicy == api.RestartPolicyAlways.

Comment (janetkuo) [inline pkg/kubectl/cmd/run.go]: Now we set `kubectl run` default to create replication controllers (`--generator=run/v1`) when restartPolicy == api.RestartPolicyAlways. We want to make it default to create deployments (`--generator=deployment/v1beta1`) in the future once Deployment reaches beta. Therefore, this TODO needs to stay.

Comment (kargakis) [inline pkg/kubectl/cmd/run.go]: You already have added it where it's supposed to be set when deployments reach beta. (L147)

Comment (janetkuo) [inline pkg/kubectl/cmd/run.go]: Yes, and this flag description needs to be updated too. I want to put this TODO in both places just in case people change the behavior without updating flag descriptions.

Comment (kargakis) [inline pkg/kubectl/cmd/run.go]: ok

**Comment by bgrant0607** [comment]
Is there anything specific left to review here?

**Comment by janetkuo** [comment]
@bgrant0607 No.

**Comment by bgrant0607** [comment]
@nikhiljindal PTAL.

File: hack/test-cmd.sh [hunk 9]
```diff
@@ -647,6 +647,24 @@ runTests() {
   # Clean up 
   kubectl delete pods test-pod "${kube_flags[@]}"
 
+  ## kubectl run should create deployements or jobs
```
Comment (nikhiljindal) [inline hack/test-cmd.sh]: deployments (typo)

File: pkg/kubectl/cmd/run.go [hunk 10]
```diff
@@ -83,7 +86,8 @@ func NewCmdRun(f *cmdutil.Factory, cmdIn io.Reader, cmdOut, cmdErr io.Writer) *c
...
+	cmd.Flags().String("generator", "", "The name of the API generator to use.  Default is 'run/v1' if --restart=Always, otherwise the default is 'job/v1beta1'.")
```
Comment (nikhiljindal) [inline pkg/kubectl/cmd/run.go]: Mention that deployment/v1beta1 is also a valid value?

File: pkg/kubectl/run.go [hunk 11]
```diff
@@ -23,10 +23,262 @@ import (
...
+
+	if err = updatePodContainers(params, args, envs, podSpec); err != nil {
+		return nil, err
+	}
+
+	if err := updatePodPorts(params, podSpec); err != nil {
+		return nil, err
+	}
+
+	deployment := extensions.Deployment{
```
Comment (nikhiljindal) [inline pkg/kubectl/run.go]: Add a TODO here specifying the issue, so that people reading this code (or copying the pattern) know that we need to fix this to generate versioned types.

**Comment by nikhiljindal** [comment]
Mostly LGTM, a few nits.

File: hack/test-cmd.sh (hunk 9 above)
Comment (janetkuo) [inline hack/test-cmd.sh]: Fixed

File: pkg/kubectl/cmd/run.go (hunk 10 above)
Comment (janetkuo) [inline pkg/kubectl/cmd/run.go]: Done

File: pkg/kubectl/run.go (hunk 11 above)
Comment (janetkuo) [inline pkg/kubectl/run.go]: Done

**Comment by janetkuo** [comment]
Fixed nits. PTAL.

**Comment by kargakis** [comment]
Needs a rebase

**Comment by janetkuo** [comment]
Rebased

**Comment by nikhiljindal** [comment]
LGTM, thanks!

**Comment by wojtek-t** [comment]
@janetkuo - can you please rebase?

**Comment by janetkuo** [comment]
Rebased

**Comment by kargakis** [comment]
@janetkuo can you please close this: https://github.com/kubernetes/kubernetes/pull/17264

**Comment by bgrant0607** [comment]
This is confusing some users, so we'll need to highlight this change.
